### PR TITLE
Decouple object safe part from GraphQLType

### PR DESCRIPTION
**Pull request by tyranron**
Revealed from #682  
Required for #682 

## Overview

Decouples object safe part of `GraphQLType` trait into a separate `GraphQLValue` trait. Does the same for `GraphQLTypeAsync` and `GraphQLSubscriptionType` too.

## Motivation

This change allows to use dynamic dispatch via resolving `GraphQLValue` trait objects.

### CODE REVIEW

**Comment by LegNeato** [comment]
Is the perf noticeable switching from static to dynamic dispatch?

**Comment by tyranron** [comment]
@LegNeato we do not switch from static to dynanic dispatch in this PR. This PR touches nothing perf-related. It just refactors traits definitions in the way making dynamic dispatch _possible_ to use in future. Current `master` doesn't allow to use dynamic dispatch at all. If you don't use dyn dispatch - you won't have it at all. And, at the moment, no code with dyn dispatch is generated, but it will be some in #682 to dispatch interfaces. I'll reason about that there.

**Comment by LegNeato** [comment]
Apologies, hadn't looked at the code yet :-)

File: juniper/src/types/containers.rs [hunk 1]
```diff
@@ -38,17 +51,17 @@ where
...
-    T: GraphQLTypeAsync<S, Context = CtxT>,
+    T: GraphQLValueAsync<S, Context = CtxT>,
     T::TypeInfo: Send + Sync,
     S: ScalarValue + Send + Sync,
     CtxT: Send + Sync,
 {
     fn resolve_async<'a>(
         &'a self,
         info: &'a Self::TypeInfo,
-        _selection_set: Option<&'a [Selection<S>]>,
```
[the remark this answers is not in the document]
Comment (LegNeato) [reply juniper/src/types/containers.rs:51]: Ok, I am convinced.

**Pull request merged**